src/librustc_typeck/check/mod.rs

```diff
@@ -3311,6 +3347,20 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             _ => self.warn_if_unreachable(expr.id, expr.span, "expression")
         }
 
+        // Non-lvalues that produce values of type `!` indicate a
```

Comment:
The lvalue thing looks like a hack. For example, isn't a dereference of an `*const !` an lvalue, as in `let x = *(ptr: *const !);`?